feat: add an encrypted value to the TransactionOutput

therustmonk opened the pull request:
Description
---
The PR adds the `encrypted_value` field (of `EncryptedValue` type) to the `TransactionOutput` struct.
The same field also added to the `TransactionInput` for a case when input was produced from an output.

The changes also affected the genesis block, dibbler's faucet data, hashes and integration tests.

The value is not actually encrypted and now it's stored in the following format: u64le + 16 empty bytes.

Motivation and Context
---
Part of the BP+ features.

How Has This Been Tested?
---
CI tests (updated)

therustmonk committed:
feat: add encrypted_field to the TransactionOutput

hansieodendaal commented on base_layer/core/src/transactions/transaction_components/encrypted_value.rs:
```diff
@@ -67,20 +65,20 @@ impl EncryptedValue {
     }
 }
 
 impl ConsensusEncoding for EncryptedValue {
```
I think  a test would be in order for the encoding/decoding

hansieodendaal commented on base_layer/core/src/transactions/transaction_components/encrypted_value.rs:
```diff
@@ -67,20 +65,20 @@ impl EncryptedValue {
-        Ok(Self { data })
+        Ok(Self(data))
     }
 }
```
I think a test would be in order for the encoding/decoding

hansieodendaal reviewed (dismissed):
LGTM

sdbondi commented on base_layer/core/src/transactions/transaction_components/encrypted_value.rs:
```diff
@@ -32,31 +32,29 @@ use crate::consensus::{ConsensusDecoding, ConsensusEncoding};
-}
+/// value: u64 + tag: [u8; 16]
+#[derive(Debug, Clone, Deserialize, Serialize, PartialEq, Eq, Hash)]
+pub struct EncryptedValue(#[serde(with = "hex::serde")] pub [u8; SIZE]);
```
Any reason we use hex to serialize this value?
This essentially doubles the storage required to 48 bytes.
You could implement the `Hex` trait from tari_utilities if you want to convert this value to and from hex

therustmonk replied:
I've added hex support to read values from the `dibbler_faucet.json` file. Ok, I'll try to make it more coherently with other types we use in the `TransactionOutput`.

sdbondi replied:
:+1: I think you may need serialize_with, something like this https://stackoverflow.com/a/67148611/1077579
Essentially, using hex only when the serialisation is "human readable"

sdbondi reviewed (commented):
Looks good - one question

hansieodendaal commented on base_layer/core/src/transactions/transaction_components/hex_patch.rs:
```diff
@@ -0,0 +1,105 @@
+use std::{fmt, marker::PhantomData};
```
The licence text is missing here

therustmonk replied:
Moved and fixed here: https://github.com/tari-project/tari_utilities/pull/43

hansieodendaal commented on base_layer/core/src/transactions/transaction_components/mod.rs:
```diff
@@ -57,6 +57,7 @@ mod committee_definition_features;
 mod encrypted_value;
 mod error;
 mod full_rewind_result;
+mod hex_patch;
```
Where is this used? Maybe it should move to `tari_utilities`?

therustmonk replied:
Yeah, I did that: https://github.com/tari-project/tari_utilities/pull/43

Commit:
Add FFI methods

- Added FFI methods to for the encrypted value
- Fixed the failing unit test
- Added new FFI tests

therustmonk committed:
Merge pull request #4 from hansieodendaal/ho_add_ffi_methods

feat: add ffi methods

sdbondi commented on base_layer/core/src/transactions/transaction_components/encrypted_value.rs:
```diff
@@ -67,20 +59,54 @@ impl EncryptedValue {
+    /// that will produce a decrypted value from self.
+    pub fn todo_decrypt(&self) -> u64 {
+        let mut buffer = [0u8; 8];
+        (&mut buffer[0..8]).copy_from_slice(&self.0.as_slice()[0..8]);
```
```suggestion
        buffer[0..8].copy_from_slice(&self.0[0..8]);
```

sdbondi reviewed (approved):
LGTM - not tested
